File: Excel/900-999/942/942 Challenge.py
import pandas as pd

path = "900-999/942/942 Matrix.xlsx"
input = pd.read_excel(path, usecols="A:B", nrows=21, skiprows=1)
test = pd.read_excel(path, usecols="D:J", nrows=6, skiprows=1)

res = input.drop_duplicates()

result = (
    input.merge(res, on="Order_ID")
       .value_counts(["Product_x", "Product_y"])
       .unstack(fill_value=0)
       .sort_index()
       .pipe(lambda x: x.reindex(sorted(x.columns), axis=1))
).reset_index()

test = test.rename(columns={"Product.1": "Product"})
result = result.rename(columns={"Product_x": "Product"})

print((result == test).all().all())

File: Excel/900-999/942/942 Challenge.R
library(tidyverse)
library(readxl)

path <- "900-999/942/942 Matrix.xlsx"
input <- read_excel(path, range = "A2:B22")
test <- read_excel(path, range = "D2:J8")

result = input %>%
  distinct(Order_ID, Product) %>%
  summarise(
    pairs = list(expand.grid(p1 = Product, p2 = Product)),
    .by = Order_ID
  ) %>%
  unnest(pairs) %>%
  count(p1, p2) %>%
  complete(p1, p2, fill = list(n = 0)) %>%
  pivot_wider(names_from = p2, values_from = n) %>%
  mutate(p1 = as.character(p1)) %>%
  arrange(p1) %>%
  select(Product = p1, sort(colnames(.)))

all.equal(result, test)
#> [1] TRUE
